xdsl/transforms/experimental/stencil_to_csl_stencil.py

```diff
@@ -0,0 +1,163 @@
+from collections.abc import Sequence
+
+from attr import dataclass
+
+from xdsl.context import MLContext
+from xdsl.dialects import memref, stencil, tensor
+from xdsl.dialects.builtin import IntegerAttr, IntegerType, ModuleOp, TensorType
+from xdsl.dialects.csl_ import csl_stencil
+from xdsl.dialects.experimental import dmp
+from xdsl.ir import Attribute, OpResult
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.hints import isa
+
+
+@dataclass(frozen=True)
+class ApplyOpAddPrefetch(RewritePattern):
+
+    prefetch_op: csl_stencil.PrefetchOp
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.ApplyOp, rewriter: PatternRewriter, /):
+
+        pass
+
+
+@dataclass(frozen=True)
+class AccessOpFromPrefetch(RewritePattern):
+
+    # index of the stencil access to be replaced by the prefetched buffer in the last arg
+    arg_index: int
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.AccessOp, rewriter: PatternRewriter, /):
+        assert len(op.offset) == 2
+        if tuple(op.offset) == (0, 0):
+            return
+        prefetched_arg = op.get_apply().region.block.args[-1]
+        assert isa(m_type := prefetched_arg.type, memref.MemRefType[Attribute])
+        assert isa(t_type := m_type.get_element_type(), TensorType[Attribute])
+
+        csl_access_op = csl_stencil.AccessOp(
+            op=prefetched_arg,
+            offset=op.offset,
+            offset_mapping=op.offset_mapping,
+            result_type=t_type,
+        )
+
+        # if the matched op is immediately followed by tensor.ExtractSliceOps to remove ghost cells, replace both
+        if len(op.res.uses) == 1 and isinstance(
+            use := list(op.res.uses)[0].operation, tensor.ExtractSliceOp
+        ):
+            rewriter.replace_op(use, csl_access_op)
+            rewriter.erase_op(op)
```

Comment:
Can I get some comments here assuring and explaining tome that this is correct? I have a hard time understanding how removing the slice op here does not invalidate the whole semantics of the code...

Reply:
It's complicated. `stencil-tensorize-z-dimension` adds extract slice ops to handle ghost cells and pass type validation. Since ghost cells are not prefetched, these ops can be removed again. If left in, they have no effect. Adding some additional checks on shape etc. to make sure it's safe to remove.

Reply:
Ok, I'll leave in the code until then if that's ok.